controlplane/internal/controller/openshiftassistedcontrolplane_controller_test.go

```diff
@@ -273,6 +274,123 @@ var _ = Describe("OpenshiftAssistedControlPlane Controller", func() {
 	})
 })
 
+var _ = Describe("Upgrade", func() {
+	Context("distributionVersion returned", func() {
+		const (
+			openshiftAssistedControlPlaneName = "test-resource"
+			clusterName                       = "test-cluster"
+			namespace                         = "test"
+		)
+		var (
+			ctx                         = context.Background()
+			cluster                     *clusterv1.Cluster
+			controllerReconciler        *OpenshiftAssistedControlPlaneReconciler
+			mockCtrl                    *gomock.Controller
+			k8sClient                   client.Client
+			mockWorkloadClientGenerator *mockWorkloadClient
+		)
+		BeforeEach(func() {
+			k8sClient = fakeclient.NewClientBuilder().
+				WithScheme(testScheme).
+				WithStatusSubresource(&controlplanev1alpha2.OpenshiftAssistedControlPlane{}).Build()
+
+			mockCtrl = gomock.NewController(GinkgoT())
+			ns := &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: namespace}}
+			Expect(k8sClient.Create(ctx, ns)).To(Succeed())
+
+			mockWorkloadClientGenerator = NewMockWorkloadClient()
+			controllerReconciler = &OpenshiftAssistedControlPlaneReconciler{
+				Client:                         k8sClient,
+				Scheme:                         k8sClient.Scheme(),
+				OpenShiftVersion:               &mockOpenShiftVersioner{},
+				WorkloadClusterClientGenerator: mockWorkloadClientGenerator,
+			}
+
+			By("creating a cluster")
+			cluster = &clusterv1.Cluster{}
+			err := k8sClient.Get(ctx, types.NamespacedName{Name: clusterName, Namespace: namespace}, cluster)
+			if err != nil && errors.IsNotFound(err) {
+				cluster = testutils.NewCluster(clusterName, namespace)
+				err := k8sClient.Create(ctx, cluster)
+				Expect(err).NotTo(HaveOccurred())
+			}
+		})
+
+		AfterEach(func() {
+			mockCtrl.Finish()
+			k8sClient = nil
+		})
+		When("the kubeconfig secret is not yet available for the workload cluster", func() {
+			It("should return an error and the distribution version should be empty", func() {
+				By("creating the openshiftassistedcontrolplane with condition kubeconfig available set to false")
+				openshiftAssistedControlPlane := getOpenshiftAssistedControlPlane()
+				conditions.MarkFalse(
+					openshiftAssistedControlPlane,
+					controlplanev1alpha2.KubeconfigAvailableCondition,
+					controlplanev1alpha2.KubeconfigUnavailableFailedReason,
+					clusterv1.ConditionSeverityInfo,
+					"error retrieving Kubeconfig %v", fmt.Errorf("secret does not exist"),
+				)
+
+				By("confirming an error is returned when calling getWorkloadClusterVersion")
+				_, err := controllerReconciler.getWorkloadClusterVersion(ctx, openshiftAssistedControlPlane)
```

Comment:
Ideally we should not test private methods. However, this doesn't necessarily belong to the controller package, so we could extract it into its own and leave the test virtually unchanged.

Reply:
Extracted it! How does it look?